### PR TITLE
Room List sorting algorithms

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/riot-web/issues/12242

image

image

when turning off the top two toggles I get:

image

### CODE REVIEW

**Comment by turt2live** [comment]
(will take a look, but I imagine it's fine - if you're confident then merge without my review)

**Comment by ara4n** [comment]
this isn't behind a labsflag, but it's still a feature we want to get right for mozilla (and others). hence asking @nadonomy to review

**Comment by aaronraimist** [comment]
To me it seems a bit confusing to have "Order rooms by message activity instead of by name" as a yes/no option.

Wouldn't it make more sense to say have a dropdown or picker? 

Order rooms by:
- Room name/alphabetical
- Message activity

That would also allow for more sorting options to be added to the list in the future

**Comment by nadonomy** [comment]
> To me it seems a bit confusing to have "Order rooms by message activity instead of by name" as a yes/no option.
> 
> Wouldn't it make more sense to say have a dropdown or picker?
> 
> Order rooms by:
> 
> * Room name/alphabetical
> * Message activity
> 
> That would also allow for more sorting options to be added to the list in the future

Yup, this is just for brevity & ease of implementation now. We have a Settings redesign in the near future where we'll move things like this to radio interactions then.

**Comment by nadonomy** [comment]
Struggling to dogfood this on https://riots.im/adhoc/alpha1/, with both of the first 2 toggles disabled, I'm seeing:

Screenshot 2020-02-21 at 15 51 45

@t3chguy any ideas?

**Comment by ara4n** [comment]
i am finding this very confusing to dogfood too. i wonder if we need some icon on the rooms to try to distinguish between rooms hoisted due to importance versus alphabet? or even a button on the section heading to toggle alpha sort on/off?

**Comment by jryans** [comment]
Hmm, this seems to work well for me when disabling both of the top 2 options so that sorting is by name and unread notifs don't affect it:

2020-02-24 at 21 37

@nadonomy, were you testing with unread notif sort (the 2nd option) off or on?

**Comment by jryans** [comment]
This blocked for the moment: @nadonomy is planning to do some more testing and tweaking of the controls.

File: src/settings/controllers/SettingController.js [hunk 1]
```diff
@@ -47,4 +47,15 @@ export default class SettingController {
...
     }
+
+    /**
+     * Gets the value which should actually get written into the store based on the input value from setValue.
+     * @param {string} level The level at which the setting has been modified.
+     * @param {String} roomId The room ID, may be null.
+     * @param {*} newValue The new value for the setting, may be null.
+     * @return {*} The value that should be used, may be null.
+     */
+    augmentValue(level, roomId, newValue) {
```
Comment (turt2live) [inline src/settings/controllers/SettingController.js]: This isn't used?

**Comment by turt2live** [comment]
also sorry about the merge conflicts

**Comment by t3chguy** [comment]
image

**Comment by t3chguy** [comment]
image

**Comment by t3chguy** [comment]
/me requests another review as Travis cleared it from the queue

**Comment by jryans** [comment]
(Cleared the product review as I assume the screenshot from @nadonomy above suffices as approval...?)

**Comment by t3chguy** [comment]
Yup

image

**Pull request merged**